clang/lib/Sema/Metafunctions.cpp

```diff
@@ -1252,6 +1257,66 @@ bool map_decl_to_entity(APValue &Result, Sema &S, EvalFn Evaluator,
   llvm_unreachable("unknown reflection kind");
 }
 
+template <typename DT>
```

Comment:
Rather than template-ize this function, we may be able to make `DT& declType` a `NamedDecl *`, as `NamedDecl` is an ancestor of `ValueDecl`, `TemplateDecl`, `NamespaceDecl`, and `TypeDecl`. Passing by reference as you've done here (i.e., `NamedDecl &`) would be just as good, but let's stick with passing by pointer since that's what's done throughout the rest of this file.

Reply:
I could but then I have to cast everything which is a bit ugly IMHO.  But then I guess it would match the rest of the clang codebase so...